mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -0,0 +1,310 @@
+        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, fnWithMeasurementsRefAttr);
+
+    // Function folded: Create the folded circuit (withoutMeasurement *
+    // Adjoint(withoutMeasurement))**scalar_factor * withMeasurements
+    rewriter.setInsertionPointToStart(moduleOp.getBody());
+    SmallVector<Type> typesFolded = {originalTypes.begin(), originalTypes.end()};
+    Type indexType = rewriter.getIndexType();
+    typesFolded.push_back(indexType);
+    FunctionType fnFoldedType = FunctionType::get(ctx, /*inputs=*/
+                                                  typesFolded,
+                                                  /*outputs=*/fnOp.getResultTypes());
+
+    func::FuncOp fnFoldedOp = rewriter.create<func::FuncOp>(loc, fnFoldedName, fnFoldedType);
+    fnFoldedOp.setPrivate();
+
+    Block *foldedBloc = fnFoldedOp.addEntryBlock();
+    rewriter.setInsertionPointToStart(foldedBloc);
+    TypedAttr numberQubitsAttr = rewriter.getI64IntegerAttr(numberQubits);
+    Value numberQubitsValue = rewriter.create<arith::ConstantOp>(loc, numberQubitsAttr);
+    Value allocQreg = rewriter.create<func::CallOp>(loc, fnAllocOp, numberQubitsValue).getResult(0);
+
+    Value c0 = rewriter.create<index::ConstantOp>(loc, 0);
+    Value c1 = rewriter.create<index::ConstantOp>(loc, 1);
+    int64_t sizeArgs = fnFoldedOp.getArguments().size();
+    Value size = fnFoldedOp.getArgument(sizeArgs - 1);
+    // Add scf for loop to create the folding
+    Value loopedQreg =
+        rewriter
+            .create<scf::ForOp>(
+                loc, c0, size, c1, /*iterArgsInit=*/allocQreg,
+                [&](OpBuilder &builder, Location loc, Value i, ValueRange iterArgs) {
+                    Value qreg = iterArgs.front();
+                    std::vector<Value> argsAndQreg = {fnFoldedOp.getArguments().begin(),
+                                                      fnFoldedOp.getArguments().end()};
+                    argsAndQreg.pop_back();
+                    argsAndQreg.push_back(qreg);
+
+                    // Call the function without measurements
+                    Value fnWithoutMeasurementsQreg =
+                        builder.create<func::CallOp>(loc, fnWithoutMeasurementsOp, argsAndQreg)
+                            .getResult(0);
+
+                    // Call the function without measurements in an adjoint region
+                    auto adjointOp = builder.create<quantum::AdjointOp>(loc, qregType,
+                                                                        fnWithoutMeasurementsQreg);
+                    Region *adjointRegion = &adjointOp.getRegion();
+                    Block *adjointBlock = builder.createBlock(adjointRegion, {}, qregType, loc);
+
+                    std::vector<Value> argsAndQregAdjoint = {fnFoldedOp.getArguments().begin(),
+                                                             fnFoldedOp.getArguments().end()};
+                    argsAndQregAdjoint.pop_back();
```

Comment:
Using `pop_back` right after the creation of `argsAndQregAdjoint` is not necessary and can be removed. :thinking:

Reply:
No here we needs to replace the last argument with the one from the adjoint block